src/asm/macro.c

```diff
@@ -40,12 +39,13 @@ struct MacroArgs *macro_NewArgs(void)
 	return args;
 }
 
-void macro_AppendArg(struct MacroArgs *args, char *s)
+void macro_AppendArg(struct MacroArgs **args, char *s)
 {
-	if (args->nbArgs == MAXMACROARGS)
+	if ((**args).nbArgs == MAXMACROARGS)
 		yyerror("A maximum of " EXPAND_AND_STR(MAXMACROARGS)
 			" arguments is allowed");
-	args->args[args->nbArgs++] = s;
+	*args = realloc(*args, sizeof **args + sizeof (char *) * (1 + (**args).nbArgs));
```

Comment:
Note that this will result in the execution time being quadratic in number of simultaneous macro arguments.

Reply:
Only if the reallocation doesn't happen in place, which is unlikely.

Reply:
That depends on how fragmented the heap is. By my measurements, the buffer still needs to be reallocated about 20% of the time when assembling a big project.